### PR TITLE
fix: retry reading query response on transaction id mismatch

### PR DESCRIPTION
**Pull request by guelfey**
This was previously attempted with the flushAll call, but this doesn't
seem to be reliable in practice - since it sets the read timeout to
time.Now(), this means that in practice the Read in most cases actually
returns directly without attempting to read.

Instead of trying to extend the deadline for the flushing read by some
arbitrary small value, this fix handles the problem at a different
place: if we detect a transaction ID mismatch by 1, we simply ignore the
response and assume that a matching response will follow. This required
some refactoring of the reading code since just retrying the read
without another write wasn't easily possible.

### CODE REVIEW

File: tcpclient_test.go [hunk 1]
```diff
@@ -93,6 +93,69 @@ func TestErrTCPHeaderLength_Error(t *testing.T) {
...
+	defer close(done)
+	data := []byte{0xCA, 0xFE}
+	go func() {
+		conn, err := ln.Accept()
+		if err != nil {
+			t.Error(err)
+			return
+		}
+		defer conn.Close()
+		time.Sleep(1 * time.Second)
```
Comment (frzifus) [inline tcpclient_test.go:113]: could this result in flacky test behavior?

**Review by frzifus — APPROVED** [review]

**Review by Carelo — APPROVED** [review]
Looks good to me, great work :+1:

File: tcpclient_test.go (hunk 1 above)
Comment (Carelo) [reply tcpclient_test.go:113]: i think this  should be fine, as long as both requests are sended before this returns.

**Review by gq0 — APPROVED** [review]
Awesome work 👍🏻 feel free to merge.
It's actually a behavior change that needs to be guarded with a feature-flag. Could we load two different versions of this lib in our client code base behind a feature flag?

**Comment by guelfey** [comment]
I see it as a bug fix actually - this was supposed to already happen

File: tcpclient_test.go (hunk 1 above)
Comment (guelfey) [reply tcpclient_test.go:113]: we have: https://buildkite.com/gridX/modbus/builds/93#ccdff30d-2ec2-47ee-a7cd-3fdd74d17ffb/207

Comment (frzifus) [reply tcpclient_test.go:113]: But so far no super slow nightly arm build or? which lets all timings exceed :dancers:

**Comment by gq0** [comment]
> I see it as a bug fix actually - this was supposed to already happen

The transaction ID handling seems new to me. But we can also roll it out slowly and see 👍🏻

**Pull request merged**